### PR TITLE
JSON Metrics integration

### PR DESCRIPTION
**Pull request by UDXS**
This PR is part of work to replace the grep-based JSON metrics extraction in OR and ORFS with JSON logging that is integrated into the App.
Currently, it lacks the `__clock__`  set of metrics.

### CODE REVIEW

**Comment by maliberty** [comment]
Why is an update of src/sta required?  You also have conflicts to resolve

File: src/utl/include/utl/Logger.h [hunk 1]
```diff
@@ -168,7 +171,7 @@ class Logger
   // API as we are writing JSON not user messages.
   // Note: these methods do no escaping so avoid special characters.
   template <typename T,
-            typename = std::enable_if_t<std::is_arithmetic_v<T>>>
+            std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
```
Comment (maliberty) [inline src/utl/include/utl/Logger.h:174]: Why this change?

File: src/utl/include/utl/Logger.h [hunk 2]
```diff
@@ -233,9 +243,15 @@ class Logger
     inline void log_metric(const std::string_view metric,
                            const Value& value)
     {
+      std::string key;
+      if (metrics_stages_.empty()) 
+        key = std::string(metric);
```
Comment (maliberty) [inline src/utl/include/utl/Logger.h:248]: Why is std::string required?  std::string::operator= should work with a std::string_view

File: src/utl/src/Logger.cpp [hunk 3]
```diff
@@ -138,4 +138,27 @@ void Logger::removeSink(spdlog::sink_ptr sink)
...
 
-}  // namespace
+void Logger::setMetricsStage(std::string_view format) {
+  popMetricsStage();
+  pushMetricsStage(format);
+}
+void Logger::clearMetricsStage() {
+  while(!metrics_stages_.empty()) {
+    metrics_stages_.pop();
+  }
```
Comment (maliberty) [inline src/utl/src/Logger.cpp:148]: You can just swap() with an empty stack.

File: src/utl/src/Logger.cpp [hunk 4]
```diff
@@ -138,4 +138,27 @@ void Logger::removeSink(spdlog::sink_ptr sink)
   }
 }
 
-}  // namespace
+void Logger::setMetricsStage(std::string_view format) {
+  popMetricsStage();
+  pushMetricsStage(format);
```
Comment (maliberty) [inline src/utl/src/Logger.cpp:143]: metric_stages_.top() = format;

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/utl/include/utl/Logger.h (hunk 1 above)
Comment (UDXS) [reply src/utl/include/utl/Logger.h:175]: The previous version did not build once the non-string `metric` was invoked. 
I think because `enable_if_t` expects a boolean from `::value` instead of the `is_arithmetic` struct itself and `enable_if_t` itself will be a `void` type if the condition is met.

**Comment by maliberty** [comment]
I'm not sure why but you have a bunch of compiler errors to resolve.

**Comment by tspyrou** [comment]
@UDXS are you stuck by anything. I can see there have been compile errors here for a few days on all but centos7.

**Comment by UDXS** [comment]
I believe I know what is causing these errors. I will commit some resolutions today.

File: src/utl/include/utl/Logger.h (hunk 1 above)
Comment (maliberty) [reply src/utl/include/utl/Logger.h:175]: It can just be changed to:

typename = std::enable_if_t<std::is_arithmetic_v<T>**::value**>

**Comment by vvbandeira** [comment]
> I believe I know what is causing these errors. I will commit some resolutions today.

@UDXS FYI, if you click on "Details" there's a tab called "Artifacts" where you can download the logs from the tests.
https://jenkins.openroad.tools/blue/organizations/jenkins/OpenROAD-Public/detail/PR-1669-merge/16/artifacts

Also, I see some metrics have non-conforming names. Will this be addressed in this or another PR?
```
    "ord_version" : ""
  , "instance_count" : "23878"
  , "floorplan__design__io" : 269
  , "timing__setup__ws" : 1.5985824e-10
  , "timing__setup__ws" : 1.5985824e-10
  , "timing__setup__tns" : 0.0
  , "timing__setup__ws" : 1.6033241e-10
  , "timing__setup__ws" : 1.6033241e-10
  , "timing__setup__tns" : 0.0
  , "utilization" : "8.4"
  , "design_area" : "59555"
  , "ANT::errors" : "0"
  , "DRT::drv" : "0"
  , "timing__setup__ws" : 8.4783514e-11
  , "timing__setup__ws" : 8.4783514e-11
  , "timing__setup__tns" : 0.0
```

**Comment by UDXS** [comment]
> Also, I see some metrics have non-conforming names. Will this be addressed in this or another PR?

They will not be addressed in this PR. There was a decision to allow metrics in flow stages that are not officially in the format (i.e `detailedroute__clock__skew__worst`) but I am not sure yet what will be done about the non-conforming metrics coming from other OR components like `ANT::errors`.

**Comment by vvbandeira** [comment]
> They will not be addressed in this PR. There was a decision to allow metrics in flow stages that are not officially in the format (i.e `detailedroute__clock__skew__worst`) but I am not sure yet what will be done about the non-conforming metrics coming from other OR components like `ANT::errors`.

I see, thanks. Was there any discussion about having the same metric reported multiple times with the same name? The example I gave before is for one design, there are six references to `timing__setup__ws` with three different values.

**Comment by UDXS** [comment]
This is the default behavior when there is no "stage" being reported (i.e. if there is no call like `utl::set_metrics_stage "detailedplace__{}"`). It should only be a problem if there is a required stage key (i.e. `placeopt__timing__setup__tns`) that is not being reported. I believe all the required stages are being reported, leaving these extraneous keys produced by miscellaneous calls to `report_metrics`. I am not sure if this needs to get addressed in this PR.

**Comment by vvbandeira** [comment]
Replaced by #2047
